src/archivematicaCommon/lib/bindpid.py

```diff
@@ -0,0 +1,459 @@
+        # param name
+        qualifier = qualified_resolve_url_param.split('_')[-1]
+        qualified_purl = get_qualified_purl(purl, qualifier)
+        template = argsdict.get(qualified_resolve_url_param)
+        if template:
+            qualified_resolve_url = _render_template(template, template_params)
+            qualified_resolve_urls.append({
+                'url': qualified_resolve_url,
+                'qualifier': qualifier
+            })
+            purl_map[qualified_purl] = qualified_resolve_url
+    return resolve_url, qualified_resolve_urls, purl_map
+
+
+def _render_request_body(argsdict, resolve_url, qualified_resolve_urls):
+    """Return a request body for the PID binding request. This is assumed to be
+    an XML document specified using a Jinja2 or Django template. It may contain
+    any of the following variables:
+    - ``naming_authority``
+    - ``pid``: the desired PID
+    - ``base_resolve_url``: where the PID's PURL should resolve to
+    - ``qualified_resolve_urls``: a list of dicts, each having ``url`` and
+      ``qualifier`` keys.
+    """
+    return _render_template(
+        argsdict['pid_request_body_template'],
+        dict(naming_authority=argsdict['naming_authority'],
+             pid=argsdict['desired_pid'],
+             base_resolve_url=resolve_url,
+             qualified_resolve_urls=qualified_resolve_urls))
+
+
+def bind_pid(**kwargs):
+    """Requests that PID ``kwargs['desired_pid']`` be created and that it
+    resolve to a specified resolve URL and possibly also one or more qualified
+    resolve URLS, as given by one or more of the ``resolve_url_template_``-type
+    keys in ``kwargs``. All args in kwargs dict; not gonna repeat it. Details:
+
+    - Performs a POST request to ``pid_web_service_endpoint`` with
+      ``pid_web_service_key`` in headers for authentication.
+    - Request body is generated from the Jinja2/Django template given in
+      ``kwargs['pid_request_body_template']
+    - A 200 status code in the response is taken to indicate that the PID
+      binding request was successful. No attempt is made to parse the response
+      body because that is not easily generalizable.
+    - Error cases observed (with IISH's handle server endpoint):
+
+      - If the PID is malformed, a response with a 500 status code will be
+        returned and the SOAP XML will contain <faultcode> and <faultstring>
+        elements.
+      - If the Payload is invalid XML, then the response will have a 400 status
+        code and the response body will be HTML.
+    """
+    _validate(kwargs)
+    resolve_url, qualified_resolve_urls, purl_map = _render_url_templates(
+        kwargs)
+    request_body = _render_request_body(kwargs, resolve_url,
+                                        qualified_resolve_urls)
+    response = requests.post(
+        kwargs['pid_web_service_endpoint'],
+        data=request_body,
+        headers={'Content-Type': 'text/xml',  # <= non-paramed assumpton alert
```

Comment:
I guess I was trying to highlight which parts of the "bind PID" request are baked-in and which are user-configurable. The fact that we have XML in the request body, the fact that we're making a POST request, etc.—these are currently non-configurable. Maybe I'll just remove it.